File: frontend/src/features/conversations/models/conversation.ts
/**
 * Conversation models and types for the frontend.
 * Based on the backend conversation schemas and API responses.
 */

import { z } from 'zod';
import { TagDenormalizedSchema } from '@/features/tags';

// Conversation status enum
export const ConversationStatus = {
  ACTIVE: 'active',
  PENDING: 'pending',
  CLOSED: 'closed',
  TRANSFERRED: 'transferred',
} as const;

export type ConversationStatusType = typeof ConversationStatus[keyof typeof ConversationStatus];

// Conversation priority enum  
export const ConversationPriority = {
  LOW: 'low',
  MEDIUM: 'medium',
  HIGH: 'high',
  URGENT: 'urgent',
} as const;

export type ConversationPriorityType = typeof ConversationPriority[keyof typeof ConversationPriority];

// Customer schema
export const CustomerSchema = z.object({
  phone: z.string(),
  name: z.string().optional(),
  email: z.string().email().optional(),
  avatar_url: z.string().url().optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
});

export type Customer = z.infer<typeof CustomerSchema>;

// Participant schema
export const ParticipantSchema = z.object({
  user_id: z.string(),
  name: z.string(),
  email: z.string().email(),
  joined_at: z.string(),
  role: z.string(),
});

export type Participant = z.infer<typeof ParticipantSchema>;

// Last message schema
export const LastMessageSchema = z.object({
  id: z.string(),
  content: z.string(),
  message_type: z.string(),
  sender_type: z.string(),
  timestamp: z.string(),
  is_read: z.boolean(),
});

export type LastMessage = z.infer<typeof LastMessageSchema>;

// Main conversation schema
export const ConversationSchema = z.object({
  _id: z.string(),
  customer: CustomerSchema.optional(),
  customer_name: z.string().nullable().optional(),
  customer_phone: z.string().optional(),
  participants: z.array(ParticipantSchema),
  status: z.nativeEnum(ConversationStatus),
  priority: z.nativeEnum(ConversationPriority),
  channel: z.string(),
  department_id: z.string().optional(),
  assigned_agent_id: z.string().optional(),
  tags: z.array(TagDenormalizedSchema).default([]),
  last_message: LastMessageSchema.optional(),
  unread_count: z.number().default(0),
  created_at: z.string(),
  updated_at: z.string(),
  closed_at: z.string().optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
});

export type Conversation = z.infer<typeof ConversationSchema>;

// Conversation list response schema (matching backend ConversationListResponse)
export const ConversationListResponseSchema = z.object({
  conversations: z.array(ConversationSchema),
  total: z.number(),
  page: z.number(),
  per_page: z.number(),
  pages: z.number(),
});

export type ConversationListResponse = z.infer<typeof ConversationListResponseSchema>;

// Conversation filters
export interface ConversationFilters {
  page?: number;
  per_page?: number;
  search?: string;
  status?: ConversationStatusType;
  priority?: ConversationPriorityType;
  channel?: string;
  department_id?: string;
  assigned_agent_id?: string;
  customer_type?: string;
  has_unread?: boolean;
  is_archived?: boolean;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

// Create conversation request
export const CreateConversationSchema = z.object({
  customer_phone: z.string(),
  customer_name: z.string().optional(),
  initial_message: z.string(),
  department_id: z.string().optional(),
  priority: z.nativeEnum(ConversationPriority).default(ConversationPriority.MEDIUM),
  tags: z.array(TagDenormalizedSchema).default([]),
  metadata: z.record(z.string(), z.unknown()).optional(),
});

export type CreateConversationRequest = z.infer<typeof CreateConversationSchema>;

// Update conversation request
export const UpdateConversationSchema = z.object({
  status: z.nativeEnum(ConversationStatus).optional(),
  priority: z.nativeEnum(ConversationPriority).optional(),
  assigned_agent_id: z.string().optional(),
  department_id: z.string().optional(),
  tags: z.array(z.string()).optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
});

export type UpdateConversationRequest = z.infer<typeof UpdateConversationSchema>;

// Conversation statistics response (matching backend ConversationStatsResponse)
export const ConversationStatsSchema = z.object({
  total_conversations: z.number(),
  active_conversations: z.number(),
  pending_conversations: z.number(),
  closed_conversations: z.number(),
  unassigned_conversations: z.number(),
  conversations_by_status: z.record(z.string(), z.number()),
  conversations_by_priority: z.record(z.string(), z.number()),
  conversations_by_channel: z.record(z.string(), z.number()),
  average_response_time_minutes: z.number(),
  average_resolution_time_minutes: z.number(),
  customer_satisfaction_rate: z.number(),
});

export type ConversationStats = z.infer<typeof ConversationStatsSchema>;

// Transfer conversation request  
export const TransferConversationSchema = z.object({
  target_department_id: z.string().optional(),
  target_agent_id: z.string().optional(),
  reason: z.string(),
  notes: z.string().optional(),
});

export type TransferConversationRequest = z.infer<typeof TransferConversationSchema>;

File: app/schemas/whatsapp/chat/conversation.py
"""Conversation-related request/response schemas."""

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from app.db.models.base import PyObjectId

# Conversation Creation
class ConversationCreate(BaseModel):
    customer_phone: str = Field(..., description="Customer phone number")
    customer_name: Optional[str] = Field(None, description="Customer name")
    customer_type: str = Field("b2c", pattern="^(b2c|b2b)$", description="Customer type")
    channel: str = Field("whatsapp", description="Communication channel")
    department_id: Optional[PyObjectId] = Field(None, description="Department ID")
    priority: str = Field("normal", pattern="^(low|normal|high|urgent)$", description="Conversation priority")
    tags: List[str] = Field(default=[], description="Initial tags")
    metadata: Dict[str, Any] = Field(default={}, description="Additional metadata")

# Conversation Update
class ConversationUpdate(BaseModel):
    customer_name: Optional[str] = Field(None, description="Updated customer name")
    priority: Optional[str] = Field(None, pattern="^(low|normal|high|urgent)$", description="Updated priority")
    department_id: Optional[PyObjectId] = Field(None, description="Updated department ID")
    metadata: Optional[Dict[str, Any]] = Field(None, description="Updated metadata")

# Conversation Transfer
class ConversationTransfer(BaseModel):
    from_agent_id: Optional[PyObjectId] = Field(None, description="Current agent ID (auto-filled if not provided)")
    to_agent_id: Optional[PyObjectId] = Field(None, description="Target agent ID")
    to_department_id: Optional[PyObjectId] = Field(None, description="Target department ID")
    reason: str = Field(..., min_length=1, max_length=500, description="Transfer reason")
    notes: Optional[str] = Field(None, max_length=1000, description="Transfer notes")

    @validator('to_agent_id', 'to_department_id')
    def validate_transfer_target(cls, v, values):
        if not v and not values.get('to_department_id') and not values.get('to_agent_id'):
            raise ValueError('Either to_agent_id or to_department_id must be provided')
        return v

# Conversation Close
class ConversationClose(BaseModel):
    reason: str = Field(..., description="Close reason")
    resolution: Optional[str] = Field(None, max_length=1000, description="Resolution details")
    customer_satisfied: Optional[bool] = Field(None, description="Customer satisfaction")
    send_survey: bool = Field(True, description="Whether to send survey to customer")
    notes: Optional[str] = Field(None, max_length=1000, description="Closing notes")

# Tag Summary for conversations
class ConversationTagSummary(BaseModel):
    id: str
    name: str
    slug: str
    display_name: Optional[str] = None
    category: str
    color: str
    usage_count: int = 0

# Conversation Response
class ConversationResponse(BaseModel):
    id: PyObjectId = Field(alias="_id")
    customer_phone: str
    customer_name: Optional[str] = None
    customer_type: str = "individual"
    status: str = "pending"
    priority: str = "normal"
    channel: str = "whatsapp"
    department_id: Optional[PyObjectId] = None
    assigned_agent_id: Optional[PyObjectId] = None
    last_message_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    archived_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    tags: List[ConversationTagSummary] = []
    message_count: int = 0
    unread_count: int = 0
    response_time_minutes: Optional[int] = None
    resolution_time_minutes: Optional[int] = None
    metadata: Dict[str, Any] = {}
    # Sentiment Analysis
    current_sentiment_emoji: Optional[str] = None
    sentiment_confidence: Optional[float] = None
    last_sentiment_analysis_at: Optional[datetime] = None

    class Config:
        populate_by_name = True
        json_encoders = {
            PyObjectId: str,
            datetime: lambda v: v.isoformat()
        }

# Conversation with Details Response
class ConversationDetailResponse(ConversationResponse):
    department_name: Optional[str] = None
    assigned_agent_name: Optional[str] = None
    transfer_history: List[Dict[str, Any]] = []
    recent_messages: List[Dict[str, Any]] = []
    notes_count: int = 0

# Conversation List Response
class ConversationListResponse(BaseModel):
    conversations: List[ConversationResponse]
    total: int
    page: int
    per_page: int
    pages: int

# Conversation Query Parameters
class ConversationQueryParams(BaseModel):
    page: int = Field(1, ge=1, description="Page number")
    per_page: int = Field(20, ge=1, le=100, description="Items per page")
    search: Optional[str] = Field(None, description="Search in customer name or phone")
    status: Optional[str] = Field(None, description="Filter by status")
    priority: Optional[str] = Field(None, description="Filter by priority")
    channel: Optional[str] = Field(None, description="Filter by channel")
    department_id: Optional[PyObjectId] = Field(None, description="Filter by department")
    assigned_agent_id: Optional[PyObjectId] = Field(None, description="Filter by assigned agent")
    customer_type: Optional[str] = Field(None, description="Filter by customer type")
    has_unread: Optional[bool] = Field(None, description="Filter by unread messages")
    created_from: Optional[datetime] = Field(None, description="Filter conversations created after this date")
    created_to: Optional[datetime] = Field(None, description="Filter conversations created before this date")
    tags: Optional[List[str]] = Field(None, description="Filter by tags")
    sort_by: str = Field("last_message_at", description="Sort field")
    sort_order: str = Field("desc", pattern="^(asc|desc)$", description="Sort order")

# Conversation Assignment
class ConversationAssignment(BaseModel):
    conversation_id: PyObjectId = Field(..., description="Conversation ID")
    agent_id: PyObjectId = Field(..., description="Agent ID to assign")
    notes: Optional[str] = Field(None, max_length=500, description="Assignment notes")

# Conversation Tag Management
class ConversationTagUpdate(BaseModel):
    conversation_id: PyObjectId = Field(..., description="Conversation ID")
    tags: List[str] = Field(..., description="Updated tags list")

# Conversation Bulk Actions
class ConversationBulkAction(BaseModel):
    conversation_ids: List[PyObjectId] = Field(..., min_items=1, max_items=100, description="List of conversation IDs")
    action: str = Field(..., pattern="^(assign|transfer|close|tag|priority)$", description="Bulk action type")
    parameters: Dict[str, Any] = Field(..., description="Action parameters")

# Conversation Statistics
class ConversationStatsResponse(BaseModel):
    total_conversations: int
    active_conversations: int
    closed_conversations: int
    unassigned_conversations: int
    conversations_by_status: Dict[str, int] = {}
    conversations_by_priority: Dict[str, int] = {}
    conversations_by_channel: Dict[str, int] = {}
    average_response_time_minutes: float = 0.0
    average_resolution_time_minutes: float = 0.0
    customer_satisfaction_rate: float = 0.0

# Conversation Activity
class ConversationActivity(BaseModel):
    conversation_id: PyObjectId = Field(..., description="Conversation ID")
    activity_type: str = Field(..., description="Activity type")
    description: str = Field(..., description="Activity description")
    metadata: Dict[str, Any] = Field(default={}, description="Activity metadata")

# SLA Status
class ConversationSLAStatus(BaseModel):
    conversation_id: PyObjectId
    response_sla_met: bool
    response_time_minutes: Optional[int]
    response_sla_threshold_minutes: int
    resolution_sla_met: bool
    resolution_time_minutes: Optional[int]
    resolution_sla_threshold_minutes: int
    escalation_required: bool 